extensions/EXT/EGL_EXT_platform_xcb.txt

```diff
@@ -0,0 +1,357 @@
+Name
+
+    EXT_platform_xcb
+
+Name Strings
+
+    EGL_EXT_platform_xcb
+
+Contributors
+
+    Yuxuan Shui <yshuiv7@gmail.com>
+
+Contacts
+
+    Yuxuan Shui <yshuiv7@gmail.com>
+
+Status
+
+    Complete
+
+Version
+
+    Version 1, 2020-08-28
+
+Number
+
+    EGL Extension #141
+
+Extension Type
+
+    EGL client extension
+
+Dependencies
+
+    Requires EGL_EXT_client_extensions to query its existence without
+    a display.
+
+    Requires EGL_EXT_platform_base.
+
+    This extension is written against the wording of version 9 of the
+    EGL_EXT_platform_base specification.
+
+Overview
+
+    This extension defines how to create EGL resources from native X11
+    resources using the functions defined by EGL_EXT_platform_base.
+
+    The native X11 resources required by this extension are xcb resources.
+    All X11 types discussed here are defined by the header `xcb.h`.
+
+New Types
+
+    None
+
+New Procedures and Functions
+
+    None
+
+New Tokens
+
+    Accepted as the <platform> argument of eglGetPlatformDisplayEXT:
+
+        EGL_PLATFORM_XCB_EXT                    0x31DC
+
+    Accepted as an attribute name in the <attrib_list> argument of
+    eglGetPlatformDisplayEXT:
+
+        EGL_PLATFORM_XCB_SCREEN_EXT             0x31DE
+
+Additions to the EGL Specification
+
+    None.
+
+New Behavior
+
+    To determine if the EGL implementation supports this extension, clients
+    should query the EGL_EXTENSIONS string of EGL_NO_DISPLAY.
+
+    This extension defines the same set of behaviors as the X11 platform
+    extension, except Xlib types are replaced with xcb types.
+
+    To obtain an EGLDisplay backed by an X11 screen, call
+    eglGetPlatformDisplayEXT with <platform> set to EGL_PLATFORM_XCB_EXT. The
+    <native_display> parameter specifies the X11 display connection to use, and
+    must point to a valid xcb `xcb_connection_t` or be NULL.  If <native_display>
+    is EGL_DEFAULT_DISPLAY, then EGL will create [1] a connection to the default
+    X11 display. The environment variable DISPLAY determines the default X11
+    display as described in the documentation of `xcb_connect`.  The value of
+    attribute EGL_PLATFORM_XCB_SCREEN_EXT specifies the X11 screen to use. If
+    the attribute is omitted from <attrib_list>, then the environment varialbe
+    DISPLAY will be used to select a screen.  Otherwise, the attribute's value
+    must be a valid screen on the display connection. If the attribute's value
+    is not a valid screen, then an EGL_BAD_ATTRIBUTE error is generated.
```

Comment:
The X11 protocol has the concept of a 'default screen', expressed through XCB as well. We should specify the following behaviour:
* if a `XCB_SCREEN` is specified, then this screen number will always be used; else
* if a `native_display` is specified, then the default screen for that display will be used; else
* if `DISPLAY` is present and contains a screen part (e.g. `:0.1` specifies a screen, `:0` does not), then that screen part will be used; else
* the default screen for that display will be used

Reply:
I'd say that the X11 protocol itself has no concept of a default screen. Only the code for parsing the `$DISPLAY` environment variable does. Also, the only way to get this default screen from xcb is via `xcb_connect`. Given an `xcb_connection_t*`, you cannot figure out the `.42` in `DISPLAY=:1.42` since it is not saved anywhere (and the env var might have changed since the connection was constructed).

Or am I missing something?

Reply:
Right you are, I apparently can't read Xlib code anymore. Which is probably a good thing. Regardless, `xcb_parse_display` forces screen 0 if no screen is explicitly specified, and I'd suggest encoding the same here, if you supply a pointer to an `xcb_connection_t` but no screen number.

Reply:
Which makes me wonder... what are the semantics of this screen number? Is it assumed that all windows are on that screen (= are a descendant of the screen's root window)? Or why does a GL implementation need to know a screen number?

Reply:
Different screens may be driven by different GPUs/different driver vendors.  It's needed for dispatching at the very least.

XCB's lack of a default screen (It has one, it just isn't available besides xcb_connect) is quite annoying IMHO.

Reply:
Going back to @fooishbar's proposal, is the best we can do with XCB then this:

* if a XCB_SCREEN is specified, then this screen number will always be used; else
* if a native_display is specified, then screen 0 will be used.
* if native_display is EGL_DEFAULT_DISPLAY, and DISPLAY is present and contains a screen part (e.g. :0.1 specifies a screen, :0 does not), then that screen part will be used; else
* screen is undefined (But AFAIK, the display creation operation will fail anyway unless EGL has a magic way to find an X server, such as assuming DISPLAY=:0.0).

Reply:
Makes sense to me.

Reply:
Maybe it's better to explicitly say the display creation operation would fail in the last case?

Reply:
I don't know.  EGL implementations have historically taken a lot of liberties here.  It seems worth leaving it undefined in case some weird use case requires the EGL library to do something clever here would be harmless.  Apps != conformance tests rarely complain about errors not showing up exactly as expected in crazy corner cases in my experience.

Reply:
OK, I will update.